snap/snapcraft.yaml

```diff
@@ -158,7 +219,15 @@ layout:
 
 apps:
   dolphin-emu:
-    command: bin/desktop-launch $SNAP/usr/local/bin/dolphin-emu
+    # extensions:
+    #   - kde-neon
+    command: usr/local/bin/dolphin-emu
+    command-chain:
+      - kf5-launch
```

Comment:
Note: I checked the difference between desktop-launch of the kde-neon extension and the kf5 launcher and the only difference I could find is that the kf5 launcher adds `"$SNAP_DESKTOP_RUNTIME/usr/sbin"` to the path. I tested desktop-launch with krita and `kde-frameworks-5-qt-5-14-core18` and everything seems to be working correctly. However, `desktop-launch` contains a number of improvements that are not present in kf5-launch, so I think we'll have more success by using desktop-launch.

But it shouldn't hurt to try both..

Reply:
Thanks for the suggestions. I've changed a lot of the dev libraries to be the non-dev versions, didn't realise what an effect it would have on the final size of the snap. With regards to the `desktop-launch` of the kde-neon extension, is that script automatically invoked when using the extension? On this branch I get an error of being unable to locate libraries unless I use the kf5 launcher. I had a look through the contents of the snap installed under `/snap` and found `/snap/dolphin-emulator/current/snap/command-chain/desktop-launch`. I'm assuming this is the desktop-launch script that comes with the kde-neon extension?

Reply:
The extension automatically adds `snap/command-chain/desktop-launch` to the command chain, but since you're not using the extension in this version, you have to add it manually to the command chain.

So remove kf5-launch from command chain, remove `bin/desktop-launch` from command and add `snap/command-chain/desktop-launch` to the command chain.

If you still get an error after that, can you pasted it here in full?

Reply:
I also found another issue; you also need to add the following in order to make `desktop-launch` work:

```yaml
environment:
  SNAP_DESKTOP_RUNTIME: $SNAP/kf5
```

Reply:
I just applied those changes, and the application runs successfully. The icons in the gtk file picker are fixed, however the theming of the actual qt application doesn't match the theming of a non-snap version such as the a self compiled binary or the flatpak. The theming can be fixed with the `qt5-gtk-platformtheme` but this brings back the original problem of missing icons in the file-picker. I'm thinking this is due to the absence of `qgnomeplatform` as per your comments on the forum post?

Theming differences (may be hard to tell but the difference is in the font styling and size)
Snap theming:
![Screenshot from 2020-07-26 13-16-27](https://user-images.githubusercontent.com/24817147/88478775-e54c5e80-cf42-11ea-9160-91b96bce48e8.png)
Flatpak/compile binary theming:
![Screenshot from 2020-07-26 13-17-17](https://user-images.githubusercontent.com/24817147/88478796-04e38700-cf43-11ea-87f6-eea066a1baf6.png)

Reply:
My attempts to add qgnomeplatform to my snap havent been successful either. I've tried copying the example that ilya fedin provided on the forum https://github.com/telegramdesktop/tdesktop/blob/b79ecb59093aa871ce47669422821df77c76421e/snap/snapcraft.yaml#L457-L482. The part fails to compile for me. I modified the part accordingly to use the qt5.14 content snap.

Reply:
I can push my changes for qgnome-platform if it would help?

Reply:
That's great, we're making progress!

I recognize the theming differences, the snap currently seems to be using the QGnomeTheme builtin to Qt itself. The Flatpak seems to be using QGnomePlatform.

For QGnomePlatform in the snap, you'll need the following parts:

```yaml
  qtconf:
    plugin: nil
    override-build: |
      KF5_DIR=/snap/kde-frameworks-5-core18-sdk/current
      install -d "$SNAPCRAFT_PART_INSTALL/usr/include/$SNAPCRAFT_ARCH_TRIPLET"
      cp -a $KF5_DIR/usr/include/$SNAPCRAFT_ARCH_TRIPLET/qt5 "$SNAPCRAFT_PART_INSTALL/usr/include/$SNAPCRAFT_ARCH_TRIPLET"
      install -d "$SNAPCRAFT_PART_INSTALL/usr/lib/$SNAPCRAFT_ARCH_TRIPLET"
      cp -a $KF5_DIR/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/qt5 "$SNAPCRAFT_PART_INSTALL/usr/lib/$SNAPCRAFT_ARCH_TRIPLET"
      cat << EOF > "$SNAPCRAFT_PART_INSTALL/qt.conf"
      [Paths]
      Prefix=$KF5_DIR/usr
      ArchData=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5
      Binaries=lib/qt5/bin
      Data=share/qt5
      Documentation=share/qt5/doc
      Examples=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/examples
      Headers=$SNAPCRAFT_STAGE/usr/include/$SNAPCRAFT_ARCH_TRIPLET/qt5
      HostBinaries=lib/qt5/bin
      HostData=$SNAPCRAFT_STAGE/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/qt5
      HostLibraries=lib/$SNAPCRAFT_ARCH_TRIPLET
      Imports=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/imports
      Libraries=lib/$SNAPCRAFT_ARCH_TRIPLET
      LibraryExecutables=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/libexec
      Plugins=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/plugins
      Qml2Imports=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/qml
      Settings=/etc/xdg
      Translations=share/qt5/translations
      EOF
    stage:
      - -usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqwayland*
      - -usr/lib/x86_64-linux-gnu/qt5/plugins/wayland*
      - -usr/lib/x86_64-linux-gnu/qt5/qml/QtWayland
    prime: [-./*]

  qgnomeplatform:
    source: https://github.com/FedoraQt/QGnomePlatform.git
    source-depth: 1
    source-tag: 0.6.0
    plugin: dump
    build-snaps:
      - kde-frameworks-5-core18-sdk
      - kde-frameworks-5-core18
    build-packages:
      - libglvnd-dev
      - libgtk-3-dev
    stage-packages:
      - libgtk-3-0
      - cinnamon-desktop-data
    override-build: |
      KF5_DIR=/snap/kde-frameworks-5-core18-sdk/current
      $KF5_DIR/usr/lib/qt5/bin/qmake -qtconf "$SNAPCRAFT_STAGE/qt.conf"
      make -j$(nproc)
      make INSTALL_ROOT="$SNAPCRAFT_PART_INSTALL/tmp" install
      cp -a "$SNAPCRAFT_PART_INSTALL/tmp/$KF5_DIR/." "$SNAPCRAFT_PART_INSTALL"
      rm -r "$SNAPCRAFT_PART_INSTALL/tmp"
    after:
      - qtconf
      - qtwayland

  qtwayland:
    source: https://github.com/qt/qtwayland.git
    source-depth: 1
    source-tag: v5.12.3
    plugin: dump
    build-snaps:
      - kde-frameworks-5-core18-sdk
      - kde-frameworks-5-core18
    build-packages:
      - libglvnd-dev
    override-build: |
      KF5_DIR=/snap/kde-frameworks-5-core18-sdk/current
      $KF5_DIR/usr/lib/qt5/bin/qmake -qtconf "$SNAPCRAFT_STAGE/qt.conf"
      make -j$(nproc)
      make INSTALL_ROOT="$SNAPCRAFT_PART_INSTALL/tmp" install
      cp -a "$SNAPCRAFT_PART_INSTALL/tmp/$KF5_DIR/." "$SNAPCRAFT_PART_INSTALL"
      cp -a "$SNAPCRAFT_PART_INSTALL/tmp/$SNAPCRAFT_STAGE/." "$SNAPCRAFT_PART_INSTALL"
      rm -r "$SNAPCRAFT_PART_INSTALL/tmp"
    after:
      - qtconf
    prime: [-./*]
```

Adding these parts should make your snap look like the Flatpak. The downside is that the size of your snap will increase a bit because it includes GTK3 in your snap.

We might be able to use the `gnome-3-28-1804` snap to reduce the size of the snap again, but I'm still figuring out how to do that exactly.

Reply:
Also; don't include `qt5-gtk-platformtheme` in your snap when you use QGnomePlatform.

Reply:
That's great, we're making progress!

I recognize the theming differences, the snap currently seems to be using the QGnomeTheme builtin to Qt itself. The Flatpak seems to be using QGnomePlatform.

For QGnomePlatform in the snap, you'll need the following parts:

*Note: don't include `qt5-gtk-platformtheme` when using QGnomePlatform.*

```yaml
  qtconf:
    plugin: nil
    override-build: |
      KF5_DIR=/snap/kde-frameworks-5-core18-sdk/current
      install -d "$SNAPCRAFT_PART_INSTALL/usr/include/$SNAPCRAFT_ARCH_TRIPLET"
      cp -a $KF5_DIR/usr/include/$SNAPCRAFT_ARCH_TRIPLET/qt5 "$SNAPCRAFT_PART_INSTALL/usr/include/$SNAPCRAFT_ARCH_TRIPLET"
      install -d "$SNAPCRAFT_PART_INSTALL/usr/lib/$SNAPCRAFT_ARCH_TRIPLET"
      cp -a $KF5_DIR/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/qt5 "$SNAPCRAFT_PART_INSTALL/usr/lib/$SNAPCRAFT_ARCH_TRIPLET"
      cat << EOF > "$SNAPCRAFT_PART_INSTALL/qt.conf"
      [Paths]
      Prefix=$KF5_DIR/usr
      ArchData=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5
      Binaries=lib/qt5/bin
      Data=share/qt5
      Documentation=share/qt5/doc
      Examples=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/examples
      Headers=$SNAPCRAFT_STAGE/usr/include/$SNAPCRAFT_ARCH_TRIPLET/qt5
      HostBinaries=lib/qt5/bin
      HostData=$SNAPCRAFT_STAGE/usr/lib/$SNAPCRAFT_ARCH_TRIPLET/qt5
      HostLibraries=lib/$SNAPCRAFT_ARCH_TRIPLET
      Imports=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/imports
      Libraries=lib/$SNAPCRAFT_ARCH_TRIPLET
      LibraryExecutables=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/libexec
      Plugins=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/plugins
      Qml2Imports=lib/$SNAPCRAFT_ARCH_TRIPLET/qt5/qml
      Settings=/etc/xdg
      Translations=share/qt5/translations
      EOF
    stage:
      - -usr/lib/x86_64-linux-gnu/qt5/plugins/platforms/libqwayland*
      - -usr/lib/x86_64-linux-gnu/qt5/plugins/wayland*
      - -usr/lib/x86_64-linux-gnu/qt5/qml/QtWayland
    prime: [-./*]

  qgnomeplatform:
    source: https://github.com/FedoraQt/QGnomePlatform.git
    source-depth: 1
    source-tag: 0.6.0
    plugin: dump
    build-snaps:
      - kde-frameworks-5-core18-sdk
      - kde-frameworks-5-core18
    build-packages:
      - libglvnd-dev
      - libgtk-3-dev
    stage-packages:
      - libgtk-3-0
      - cinnamon-desktop-data
    override-build: |
      KF5_DIR=/snap/kde-frameworks-5-core18-sdk/current
      $KF5_DIR/usr/lib/qt5/bin/qmake -qtconf "$SNAPCRAFT_STAGE/qt.conf"
      make -j$(nproc)
      make INSTALL_ROOT="$SNAPCRAFT_PART_INSTALL/tmp" install
      cp -a "$SNAPCRAFT_PART_INSTALL/tmp/$KF5_DIR/." "$SNAPCRAFT_PART_INSTALL"
      rm -r "$SNAPCRAFT_PART_INSTALL/tmp"
    after:
      - qtconf
      - qtwayland

  qtwayland:
    source: https://github.com/qt/qtwayland.git
    source-depth: 1
    source-tag: v5.14.2
    plugin: dump
    build-snaps:
      - kde-frameworks-5-core18-sdk
      - kde-frameworks-5-core18
    build-packages:
      - libglvnd-dev
    override-build: |
      KF5_DIR=/snap/kde-frameworks-5-core18-sdk/current
      $KF5_DIR/usr/lib/qt5/bin/qmake -qtconf "$SNAPCRAFT_STAGE/qt.conf"
      make -j$(nproc)
      make INSTALL_ROOT="$SNAPCRAFT_PART_INSTALL/tmp" install
      cp -a "$SNAPCRAFT_PART_INSTALL/tmp/$KF5_DIR/." "$SNAPCRAFT_PART_INSTALL"
      cp -a "$SNAPCRAFT_PART_INSTALL/tmp/$SNAPCRAFT_STAGE/." "$SNAPCRAFT_PART_INSTALL"
      rm -r "$SNAPCRAFT_PART_INSTALL/tmp"
    after:
      - qtconf
    prime: [-./*]
```

Adding these parts should make your snap look like the Flatpak. The downside is that the size of your snap will increase a bit because it includes GTK3 in your snap.

We might be able to use the `gnome-3-28-1804` snap to reduce the size of the snap again, but I'm still figuring out how to do that exactly.

The only possible issue I see is that you're also including `libgtk2.0-0` in the snap. Mixing GTK 2 and GTK 3 might cause issues. Does Dolphin require GTK 2?

Reply:
Note: I edited my previous message to use `kde-frameworks-5-qt-5-14-core18` instead of `kde-frameworks-5-core18`.

Reply:
Ah I was missing the qtwayland part. I've now managed to get qgnomeplatform included within the snap, which does solve the theming problems, however it has reintroduced the missing icons in the file-picker. I didn't notice any increase in the final snap size. I have pushed the qgnomeplatform changes to the branch if you want to take a look.

I thought GTK 2 was a dependency according to the dolphin-emu build instructions, but I tested the snap without it staged and it ran successfully so it was probably only a build time dependency.